frontend/app/lib/server/open-payments.server.ts

```diff
@@ -0,0 +1,319 @@
+  const quote = await opClient.quote
+    .create(
+      {
+        url: new URL(senderAddress.id).origin,
+        accessToken: quoteGrant.access_token.value
+      },
+      {
+        method: 'ilp',
+        walletAddress: senderAddress.id,
+        receiver: incomingPayment.id,
+        debitAmount: amountObj
+      }
+    )
+    .catch((error) => {
+      console.log({ error })
+      throw new Error(
+        `Could not create quote for receiver ${receiverAddress.publicName}.`
+      )
+    })
+
+  const response = {
+    incomingPaymentGrantToken: incomingPaymentGrant.access_token.value,
+    ...quote
+  }
+
+  return response
+}
+
+type CreateIncomingPaymentParams = {
+  accessToken: string
+  walletAddress: WalletAddress
+  note: string
+  opClient: AuthenticatedClient
+}
+
+async function createIncomingPayment({
+  accessToken,
+  walletAddress,
+  note,
+  opClient
+}: CreateIncomingPaymentParams) {
+  // create incoming payment without amount
+  return await opClient.incomingPayment
+    .create(
+      {
+        url: new URL(walletAddress.id).origin,
+        accessToken: accessToken
+      },
+      {
+        expiresAt: new Date(Date.now() + 6000 * 60).toISOString(),
+        walletAddress: walletAddress.id,
+        metadata: {
+          description: note
+        }
+      }
+    )
+    .catch(() => {
+      throw new Error('Unable to create incoming payment.')
+    })
+}
+
+export async function createInteractiveGrant(args: {
+  walletAddress: WalletAddress
+  redirectUrl?: string
+}) {
+  const opClient = await createClient()
+  const clientNonce = crypto.randomUUID()
+  const paymentId = createId()
+
+  const outgoingPaymentGrant = await createOutgoingPaymentGrant({
+    walletAddress: args.walletAddress,
+    debitAmount: {
+      value: String(1 * 10 ** args.walletAddress.assetScale),
+      assetCode: args.walletAddress.assetCode,
+      assetScale: args.walletAddress.assetScale
+    },
+    nonce: clientNonce,
+    paymentId: paymentId,
+    opClient,
+    redirectUrl: args.redirectUrl
+  })
+
+  return outgoingPaymentGrant
+}
+
+export interface Amount {
+  value: string
+  assetCode: string
+  assetScale: number
+}
+
+type CreateOutgoingPaymentParams = {
+  walletAddress: WalletAddress
+  debitAmount?: Amount
+  receiveAmount?: Amount
+  nonce?: string
+  paymentId: string
+  opClient: AuthenticatedClient
+}
+
+async function createOutgoingPaymentGrant(
+  params: CreateOutgoingPaymentParams & { redirectUrl?: string }
+): Promise<PendingGrant> {
+  const {
+    walletAddress,
+    debitAmount,
+    receiveAmount,
+    nonce,
+    paymentId,
+    opClient,
+    redirectUrl
+  } = params
+
+  const grant = await opClient.grant
+    .request(
+      {
+        url: walletAddress.authServer
+      },
+      {
+        access_token: {
+          access: [
+            {
+              identifier: walletAddress.id,
+              type: 'outgoing-payment',
+              actions: ['create', 'read', 'list'],
```

Comment:
This might be enough? Mostly, `'list'` be unnecessary.
```suggestion
              actions: ['create'],
```